src/explore-education-statistics-common/src/modules/table-tool/components/workers/downloadTable.worker.ts

```diff
@@ -0,0 +1,15 @@
+import 'core-js/stable'; // required for IE 11 support
+import 'regenerator-runtime/runtime'; // required for IE 11 support
+import { getCsvData } from '@common/modules/table-tool/components/utils/downloadCsvUtils';
+
+/* eslint-disable no-restricted-globals */
+const ctx: Worker = self as never;
+
+/**
+ * Process the csv data in a new thread and post the result back to the main thread.
+ */
+ctx.addEventListener('message', event => {
+  const result = getCsvData(event.data.fullTable);
```

Comment:
Ah! Think we might have fallen into one of the main traps of using web workers. When passing data between to/from web workers, JS classes lose all of their inheritance/type information as it gets serialized/deserialized using a structured clone algorithm: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm

This means, you're most likely just working with plain JS objects within the worker itself. Whilst this might not be an immediate problem - the main issue is that the types are completely wrong and could cause issues if we want to do things like `instanceof` checks or use the `id` getter. We might just have been fortunate enough that the code doesn't do any of that so far.

I think there's two main fixes. We either:

1. Change the type of the `fullTable` argument that is received by `getCsvData` to the correct version. This isn't super ideal as the type will lack information about what the filter types are for example.

2. Re-work all of the `mapFullTable` stuff so that we no longer use classes, but change everything to plain JS objects that attach their type information instead.

With option 2 - we would swap the classes for interfaces that would look something like:
```ts
interface Filter {
  value: string;
  label: string;
  group: string;
  id: string;
}

interface CategoryFilter extends Filter {
  isTotal: boolean;
  category: string;
  type: 'category';
}

interface LocationFilter extends Filter {
  level: boolean;
  geoJson?: GeoJsonFeature[];
  type: 'location';
}

// etc etc
```

If I could re-write table tool, option 2 is what I would go with :wink: It would make it pretty simple to move data to/from workers without needing any additional re-mapping to/from classes.

Understand that this might be a little daunting though, so fine with going with option 1 for now. If we start doing more web worker stuff in the future, I think we'll want to consider option 2 more seriously.

Reply:
I wasn't aware of this, thanks! Unsurprisingly I don't really want to go down the re-working mapFullTable route right now :)
For option 1, I'm not sure what you mean by changing the argument to the correct version, can you expand on that a bit?

Reply:
No problem, suspected as much :laughing:

By having a correct version I mean we need some kind of type, maybe called `RawFullTable`, that is basically the same as `FullTable` (and `FullTableMeta`), but with the actual correct typings. It might look something like this:

```ts
// rawFullTable.ts

interface RawFilter {
  value: string;
  label: string;
  group?: string;
  id: string // Not entirely sure this would get serialized into the worker properly - please check
}

interface RawCategoryFilter extends RawFilter {
  isTotal: boolean;
  category: string;
}

// the other filter types...

export interface RawFullTableMeta {
  publicationName: string;
  subjectName: string;
  filters: Dictionary<{
    name: string;
    options: RawCategoryFilter[];
  }>;
}

export interface RawFullTable {
  title?: string;
  subjectMeta: RawFullTableMeta;
  results: TableDataResult[];
}
```

We would then use `RawFullTable` as the argument for `getCsvData` instead of `FullTable`.

I think we would want to preface this entire file by saying that we should only be using this in web worker contexts currently, and ideally we want to refactor `mapFullTable` to avoid classes. Could you ticket that up as well?

Reply:
Thanks. I'm doing this now and the id is not being serialised into the worker. The code doesn't use the id at all. so could just leave it like that (with a comment explaining), but it would be better to sort it out. I could clone the fulltable and loop over it to convert the ids to strings, or can you think of a better way?

Reply:
Ok, I've left it without any more transforms and with comments about not using the id.